pkg/labels/labels.go

```diff
@@ -32,82 +32,74 @@ func HasLabelWithValue(component Labeled, labelKey string, labelValue string) bo
 	return GetLabel(component, labelKey) == labelValue
 }
 
-// SetWatchedByK8ssandraCluster sets the required labels for making a component watched by the K8ssandraCluster, i.e. a
-// modification of the component will trigger a reconciliation loop in K8ssandraClusterReconciler.
-// klusterKey specifies the namespace and name of the K8ssandraCluster.
-func SetWatchedByK8ssandraCluster(component Labeled, klusterKey client.ObjectKey) {
-	AddLabel(component, k8ssandraapi.K8ssandraClusterNameLabel, klusterKey.Name)
-	AddLabel(component, k8ssandraapi.K8ssandraClusterNamespaceLabel, klusterKey.Namespace)
+// LabelSet represents a set of labels keys and values that are applied to a K8ssandra component.
+type LabelSet map[string]string
+
+func newLabelSet(m map[string]string) *LabelSet {
+	l := LabelSet(m)
+	return &l
 }
 
-// IsWatchedByK8ssandraCluster checks whether the given component is watched by a K8ssandraCluster.
-// klusterKey specifies the namespace and name of the K8ssandraCluster.
-func IsWatchedByK8ssandraCluster(component Labeled, klusterKey client.ObjectKey) bool {
-	return HasLabelWithValue(component, k8ssandraapi.K8ssandraClusterNameLabel, klusterKey.Name) &&
-		HasLabelWithValue(component, k8ssandraapi.K8ssandraClusterNamespaceLabel, klusterKey.Namespace)
+// ApplyTo applies the labels to the given component.
+func (l *LabelSet) ApplyTo(component Labeled) {
+	component.SetLabels(*l)
 }
 
-// WatchedByK8ssandraClusterLabels returns the labels used to make a component watched by a K8ssandraCluster.
-// klusterKey specifies the namespace and name of the K8ssandraCluster.
-func WatchedByK8ssandraClusterLabels(klusterKey client.ObjectKey) map[string]string {
-	return map[string]string{
-		k8ssandraapi.K8ssandraClusterNameLabel:      klusterKey.Name,
-		k8ssandraapi.K8ssandraClusterNamespaceLabel: klusterKey.Namespace,
+// IsPresent checks whether the labels are present on the given component.
+func (l *LabelSet) IsPresent(component Labeled) bool {
+	for key, value := range *l {
+		if !HasLabelWithValue(component, key, value) {
+			return false
+		}
 	}
+	return true
 }
 
-// SetReplicatedBy sets the required labels that make a Secret selectable by a ReplicatedSecret for a given
-// K8ssandraCluster.
-// klusterKey specifies the namespace and name of the K8ssandraCluster.
-func SetReplicatedBy(component Labeled, klusterKey client.ObjectKey) {
-	AddLabel(component, k8ssandraapi.ReplicatedByLabel, k8ssandraapi.ReplicatedByLabelValue)
-	AddLabel(component, k8ssandraapi.K8ssandraClusterNameLabel, klusterKey.Name)
-	AddLabel(component, k8ssandraapi.K8ssandraClusterNamespaceLabel, klusterKey.Namespace)
+// AsMap returns the labels as a key/value map.
+// DO NOT MODIFY THE RETURNED MAP. For performance reasons, it is shared with the internal representation.
+func (l *LabelSet) AsMap() map[string]string {
+	return *l
```

Comment:
Trying to avoid extra allocations, although this means we'll have to be disciplined about not mutating the map.